clang/lib/Sema/SemaSYCL.cpp

```diff
@@ -3309,6 +3316,25 @@ void Sema::MarkDevice(void) {
           }
           break;
         }
+        case attr::Kind::SYCLIntelFPGAMaxConcurrency: {
+          auto *SIMCA = cast<SYCLIntelFPGAMaxConcurrencyAttr>(A);
+          if (auto *Existing =
+                  SYCLKernel->getAttr<SYCLIntelFPGAMaxConcurrencyAttr>()) {
+            ASTContext &Ctx = getASTContext();
+            if (Existing->getNThreadsExpr() > SIMCA->getNThreadsExpr()) {
+              Diag(SYCLKernel->getLocation(),
+                   diag::err_conflicting_sycl_kernel_attributes);
+              Diag(Existing->getLocation(), diag::note_conflicting_attribute);
+              Diag(SIMCA->getLocation(), diag::note_conflicting_attribute);
+              SYCLKernel->setInvalidDecl();
+            } else {
+              SYCLKernel->addAttr(A);
+            }
+          } else {
+            SYCLKernel->addAttr(A);
+          }
+          break;
+        }
```

Comment:
What is the purpose adding this diagnostic here?

Reply:
I added to be able to check on things like this, previous attribute on a function.
[[intel::max_concurrency(4)]] void func();
[[intel::max_concurrency(5)]] void func() {}
I need to add a test case for this.

Reply:
You need a "merge" function to catch this:
 SYCLIntelFPGAMaxConcurrencyAttr *Sema::MergeSSYCLIntelFPGAMaxConcurrencyAttr(
    Decl *D, const SYCLIntelFPGAMaxConcurrencyAttr &A) {

Reply:
Ok. Studying that! Thanks.

Reply:
I think adding the code here only handles diagnostics when applying attributes to SYCL kernel i.e. when attributes are propagated to kernel from device functions called by kernel. I don't think duplicate attributes on any other function will be handled by this

Based on code in L563, it looks like while the attribute can be explicitly specified on kernel, it is not propagated to kernel from device functions. In this case, I do not think this block of code is required. I think the "merge" function as @smanna12 should cover kernel function as well. However please verify with tests for 'normal' and kernel functions.